Fix partial renderer assuming `locals` responds to `symbolize_keys`

paracycle opened the pull request:
### Summary

Since https://github.com/rails/rails/pull/30647 was merged, partial renderer does not play well with anything that is passed to `locals` which does not respond to `symbolize_keys`.

Before the PR, the `locals` option value was passed along untouched, or was substituted with an emtpy hash if it wasn't supplied. The PR added an operation to symbolize keys of the passed object which ends up making an assumption that the option value will always be a Hash.

There are legitimate cases, however, where a Hash-like object could be supplied instead, like an `ActionController::Parameters` instance, which does not respond to `symbolize_keys`.

This PR reverts the changes made in the previous PR to restore the use of symbols as the only supported method of passing in values.

An extra test is included that explicitly checks the case where a Hash-like object is supplied as the `locals` option.

gmcgibbon commented on actionview/lib/action_view/renderer/partial_renderer.rb:
```diff
@@ -363,7 +363,8 @@ def setup(context, options, block)
 
-        @locals  = options[:locals] ? options[:locals].symbolize_keys : {}
+        @locals  = options[:locals] || {}
+        @locals.symbolize_keys! if @locals.respond_to?(:symbolize_keys!)
```
Does it make more sense to do `options[:locals].to_h.symbolize_keys` instead?

paracycle replied:
I've considered that also, but that has other side-effects on the object that was passed in as well.

Consider this:
```ruby
render partial: "customer", locals: ActionController::Parameters.new(customer: { name: 'David' }).permit!
```
if we were to call `.to_h.symbolize_keys` on the value that is passed it, `to_h` will convert it to `{"customer"=>{"name"=>"David"}}` and `symbolize_keys` will then convert it to `{customer: {"name"=>"David"}}`.

However, if the application was expecting the name using `:name` it would break.

Passing the object untouched allows us to keep the old behaviour. In the case of `AC:Parameters`, since it allows indifferent access, `:name` works if nothing messes with it.

rafaelfranca commented:
To be fair I prefer to revert the PR that introduced this code. I find the usage of both symbols and string keys confusing and inconsistent. Most of the places we only support symbols and opening an exception here will just cause more PRs adding support in other places.

Can you change the PR to revert that commit?

paracycle committed:
Revert "Allow usage of strings as locals for partial renderer"

paracycle commented:
@rafaelfranca Done!

paracycle committed:
Add test for Hash-like object being passed to partial `locals`

The test passes an instance of `ActionController::Parameters` that acts
like a Hash but does not respond to some Hash methods like
`symbolize_keys`.

Moreover, if someone were to call `to_h` on the value it would fail since
the parameter is not permitted. So this is a great way to ensure that the
partial rendering pipeline does not mess with `locals`.
